modules/packages/OrderedSet/Treap.chpl

```diff
@@ -0,0 +1,872 @@
+      var result = false;
+
+      on this {
+        _enter();
+        result = _remove(_root, x);
+        _leave();
+      }
+
+      return result;
+    }
+
+    /*
+      Clear the contents of this orderedSet.
+
+      .. warning::
+
+        Clearing the contents of this orderedSet will invalidate all existing
+        references to the elements contained in this orderedSet.
+    */
+    proc ref clear() {
+      on this {
+        _enter();
+        if _root != nil {
+          // Delete the tree recursively by nodes' deinit
+          delete _root;
+          _root = nil;
+        }
+        _leave();
+      }
+    }
+
+    /*
+      Helper procedure to return predecessor/successor of one node
+      if direction is 0, return predecessor
+      else if 1, return successor
+    */
+    pragma "no doc"
+    proc _neighbour(in node: nodeType, in direction: int) {
+      // Assuming direction is 1, we're finding the successor
+      if node == nil then return nil;
+      if node!.children[direction] {
+        // node has right child,
+        // find the leftmost node in its right child tree 
+        node = node!.children[direction];
+        direction ^= 1;
+        while node != nil && node!.children[direction] != nil {
+          node = node!.children[direction];
+        }
+        return node;
+      }
+      else {
+        // node doesn't have right child,
+        // find its first ancesstor whose left child tree it belongs to
+        direction ^= 1;
+        while node!.parent != nil && node!.parent!.children[direction] != node {
+          node = node!.parent;
+        }
+        return node!.parent;
+      }
+    }
+
+    pragma "no doc"
+    proc const _lower_bound(node: nodeType, e: eltType): nodeType {
+      if node == nil then return nil;
+      var cmp = _compare(e, node!.element);
+      if cmp == 0 then return node;
+      else if cmp < 0 {
+        var result = _lower_bound(node!.children[0], e);
+        if result != nil then return result;
+        else return node;
+      }
+      else return _lower_bound(node!.children[1], e);
+    }
+
+    pragma "no doc"
+    proc const _upper_bound(node: nodeType, e: eltType): nodeType {
+      if node == nil then return nil;
+      var cmp = _compare(e, node!.element);
+      if cmp >= 0 then return _upper_bound(node!.children[1], e);
+      else {
+        var result = _upper_bound(node!.children[0], e);
+        if result != nil then return result;
+        else return node;
+      }
+    }
+
+    /*
+      Find the first element in the orderedSet
+      which does not compare less than e.
+
+      Returns a tuple containing two element:
+      The first element is a `bool` that indicates whether there is such an element.
+      The second element is the occurence in the orderedSet, if there's any.
+
+      :returns: a tuple containing result
+      :rtype: `(bool, eltType)`
+    */
+    proc const lowerBound(e: eltType): (bool, eltType) {
```

Comment:
The iteration can be done with `successor`. Actually I propose `successor/predecessor` for this kind of usages because Chapel don't have iterator operations like C++.

Reply:
Note iterating from lowerBound to upperBound is not really meaningful for OrderedSet. There are unique elements so only one element will be iterated in your example. This is more like a search operation.

Reply:
Note in the example, the intended iteration is from `lower_bound(10)` to `upper_bound(100)`.  Even with unique elements there are potentially 90 elements to iterate over (with `defaultComparator`.)